src/UCTNode.cpp

```diff
@@ -291,6 +296,8 @@ UCTNode* UCTNode::uct_select_child(int color, bool is_root) {
         }
     }
 
+    // the inflate() method has to be protected by the parent's lock
```

Comment:
Oh, right, this....

What was the thread restriction on inflate? If it's just to avoid concurrent inflates, we're fine. But if there's restrictions in reading out the node info while it's being inflated, this might not work. But I think we replace the data item in an atomic transaction, so it should work?

Reply:
Oh right... this won't work.  Thread-safe UCTNodePointer was #1179 and it was dropped.  See methods like UCTNodePointer::get_policy() and you can see nothing is atomic here.

So, it's either we revive #1179 or revive #1725 or find something else or just give up.  I think #1725 is simpler than #1179 as long as the RWLock has no bugs.  read-write locks are well understood concepts...

One thing is that I was running with this patch enabled for at least 40 hours and nothing crashed.  Maybe there is something that prevents any obvious crash?

Reply:
>Oh right... this won't work. Thread-safe UCTNodePointer was #1179 and it was dropped. See methods like UCTNodePointer::get_policy() and you can see nothing is atomic here.
...
>One thing is that I was running with this patch enabled for at least 40 hours and nothing crashed. Maybe there is something that prevents any obvious crash?

m_data is a 64-bit integer. On x86/x64 it will get atomically updated (the entire cache line, really), just the ordering isn't guaranteed (but by default x86 does very sane things). But nothing is going to see a half-updated value...

We can revive #1179. I closed it because there was no gain at the time and we found the bug that prompted its creation.

We can also do the RWLock. I'm not sure yet what the best approach is. But I think with very high concurrency (batching?), the atomic approach avoids taking any locks at all, and was little write traffic, whereas the RWLocks still need to write something every time there is a reader.

Reply:
On the other hand, the RWLock approach splits up the complexity into the RWLock implementation. Reasoning about the tree is probably easier with RWLocks.

Could we test how both approaches behave?